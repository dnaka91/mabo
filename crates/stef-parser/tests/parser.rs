use std::{
    fmt::{self, Display},
    fs,
};

use insta::{assert_snapshot, glob, with_settings};
use miette::{Diagnostic, MietteHandler, MietteHandlerOpts, ReportHandler};
use stef_parser::Schema;

#[test]
fn parse_schema() {
    glob!("inputs/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str()).unwrap();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("parse", format!("{value:#?}"));
            assert_snapshot!("print", value.to_string());
        });
    });
}

#[test]
fn parse_invalid_schema() {
    struct Wrapper<'a>(&'a MietteHandler, &'a dyn Diagnostic);

    impl<'a> Display for Wrapper<'a> {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            self.0.debug(self.1, f)
        }
    }

    let handler = MietteHandlerOpts::new()
        .color(false)
        .terminal_links(false)
        .width(120)
        .force_graphical(true)
        .build();

    glob!("inputs/invalid/*.stef", |path| {
        let input = fs::read_to_string(path).unwrap();
        let value = Schema::parse(input.as_str()).unwrap_err();

        with_settings!({
            description => input.trim(),
            omit_expression => true,
        }, {
            assert_snapshot!("error", Wrapper(&handler, &*value).to_string());
        });
    });
}
