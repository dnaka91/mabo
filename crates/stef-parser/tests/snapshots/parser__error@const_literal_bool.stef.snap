---
source: crates/stef-parser/tests/parser.rs
description: "const VALUE: bool = truze;"
input_file: crates/stef-parser/tests/inputs/invalid/const_literal_bool.stef
---
stef::parse::const_def (https://docs.rs/stef-parser/0.1.0/stef_parser/error/struct.ParseConstError.html)

  × Failed to parse const declaration
  ├─▶ Failed to parse literal value
  ╰─▶ error Tag
   ╭────
 1 │ const VALUE: bool = truze;
   · ─────────────┬────────────
   ·              ╰── In this declaration
   ╰────
  help: Expected const declaration in the form `❬B❭const <NAME>: <type> = <literal>;❬B❭`

Error: stef::parse::literal (https://docs.rs/stef-parser/0.1.0/stef_parser/error/struct.ParseLiteralError.html)

  × Failed to parse literal value
  ╰─▶ error Tag
   ╭────
 1 │ const VALUE: bool = truze;
   ·                     ▲
   ·                     ╰── In this declaration
   ╰────
  help: Expected literal value declaration in either of the forms:
        `❬B❭true❬B❭` or `❬B❭false❬B❭` for booleans
        `❬B❭1, 2, 3, ...❬B❭` for numbers
        `❬B❭1.2, 1.0e5, ...❬B❭` for floating point numbers
        `❬B❭"..."❬B❭` for strings
        or `❬B❭[...]❬B❭` for bytes

Error:   × error Tag

